### DISCUSSION TITLE
Upgrade to 8.10.2

### DISCUSSION THREAD

**Discussion by justindthomas** [zwave-js/zwave-js-ui#2164] [Q&A]
How long should the configuration upgrade take? I clicked the button to launch it when I upgraded zwavejs2mqtt earlier this afternoon, but it's still got the red exclamation point telling me it needs to upgrade.

When I tell it to go again, the log file complains that it's already in progress. It's been about 4 hours or so.

~~~~~
2022-01-11T02:43:05.666Z DRIVER   Installing version 8.10.2-20220110 of configuration DB...
2022-01-11T02:43:05.787Z DRIVER   Config update failed: Another installation is already in progress! (ZW0345)
~~~~~

I'm wondering if there's a lock file that I need to remove and retry.

**Comment by robertsLando** [comment]
@justindthomas Could you add zwavejs logs? 

cc @AlCalzone

**Comment by AlCalzone** [comment]
Installing the configuration fails when zwave-js isn't able to acquire the lockfile `package.json.lock` which should be next to `node_modules/zwave-js/package.json`. @justindthomas how are you running z2m? Not via Docker it seems?

**Reply by justindthomas** [reply]
I'm running it bare metal on a dedicated Raspberry Pi 4. I set it up this way just in the last few days so that I could track development more closely (I was running it in HassOS previously).

Any likely harm in me manually removing that lock file? I don't see anything in the logs to indicate there's an ongoing process that's using it, other than those two messages when I attempt the update. If there's doubt, I can attach the logs per @robertsLando question above, but they're pretty big.

**Reply by AlCalzone** [reply]
If its still there, just delete it. Normally the lockfile library should take care of that, maybe something changed the permissions so it can't.

**Reply by justindthomas** [reply]
I think I found the issue. I had decided to make use of the `package.sh` script to separate the running service from the code base (e.g., to compile it in one place, package it, and move the packaged file to another place where it runs continuously). I suspect the lock file got bundled in to that package.

When I moved the `store` folder to where the code is and ran it there using `yarn start` I was able to update the configuration DB successfully.

EDIT: Or maybe it's more complicated. I don't see any reference to `package.json.lock` in a  `strings zwavejs2mqtt` of the packaged binary. Since the lock is intended to be in the packaged  `node_modules` folder, is it just not possible to update the configuration DB from the packaged executable?

**Reply by robertsLando** [reply]
>  is it just not possible to update the configuration DB from the packaged executable?

I never tried it, BTW in packaged executable what's inside snapshot is readonly so I think it shouldn't work

**Reply by AlCalzone** [reply]
If the packaged executable is using the external configuration directory, zwave-js creates a lockfile `config-update.lock` in that directory and updates only files there.

**Reply by robertsLando** [reply]
> If the packaged executable is using the external configuration directory, zwave-js creates a lockfile config-update.lock in that directory and updates only files there.

Ok if so it should work, he needs to set the env var to use the external configuration directory as we are doing on docker now: https://github.com/zwave-js/zwavejs2mqtt/blob/master/docker/Dockerfile#L62

@justindthomas try to set this env var pointing to a directory on store or where you prefer:

```
ZWAVEJS_EXTERNAL_CONFIG=/usr/src/app/store/.config-db
```

**Reply by justindthomas** [reply]
Thanks @robertsLando. That seems to do what we expect, but I do get these errors on startup:

~~~~~
ZWAVEJS_EXTERNAL_CONFIG=/srv/zwave/.config-db ZWAVEJS_LOG_DIR=/ramdisk/logs ./zwavejs2mqtt
~~~~~

~~~~~
2022-01-12T21:51:30.931Z CONFIG   Synchronizing external config dir /srv/zwave/.config-db...
2022-01-12T21:51:30.942Z CONFIG   Synchronizing external config dir failed - using embedded config
2022-01-12T21:51:30.943Z CONFIG   version 8.10.2-20220110
~~~~~

Is that working as expected? It created the `.config-db` directory, but did not populate it.

I can't test the upgrade right now just because I've already run it from the non-packaged directory.

**Reply by robertsLando** [reply]
>  but I do get these errors on startup:

Based on what I know is that zwavejs will copy files there once you do an update of config db. @AlCalzone right?

**Reply by AlCalzone** [reply]
> Is that working as expected?

Nope. Unfortunately the error isn't logged. Can you edit the file `node_modules/@zwave-js/config/build/utils.js`?
Right at the end:
```diff
-    catch {
+    catch (e) {
        // Something went wrong
-        logger.print(`Synchronizing external config dir failed - using embedded config`, "error");
+        logger.print(`Synchronizing external config dir failed - using embedded config. Error: ${e.stack}`, "error");
```
That should show us why

**Reply by justindthomas** [reply]
Okay, great. Here's the log (and a little bit more surrounding context since I noticed a subsequent error) with that change.

~~~~~
2022-01-13T17:17:35.109Z DRIVER   loading configuration...
2022-01-13T17:17:35.114Z CONFIG   Synchronizing external config dir /srv/zwave/.config-db...
2022-01-13T17:17:35.124Z CONFIG   Synchronizing external config dir failed - using embedded config. Error: Error
                                  : ENOENT: no such file or directory, copyfile '/snapshot/zwavejs2mqtt/node_mod
                                  ules/@zwave-js/config/config/sensorTypes.json' -> '/srv/zwave/.config-db/senso
                                  rTypes.json'
2022-01-13T17:17:35.126Z CONFIG   version 8.10.2
2022-01-13T17:17:42.541Z CONFIG   Writing the device index to disk failed: File or directory '/**/zwavejs2mqtt/n
                                  ode_modules/@zwave-js/config/config/devices/index.json' was not included into 
                                  executable at compilation stage. Please recompile adding it as asset or script
                                  .
2022-01-13T17:17:42.543Z CONFIG   Priority device configuration directory /home/ubuntu/code/zwavejs2mqtt/store/c
                                  onfig not found
2022-01-13T17:17:42.590Z DRIVER   beginning interview...
~~~~~

I did check the `package.json` and I see that `node_modules/@zwave-js/config/config/devices/**` is included in the `pkg` section. I tried making that broader by dropping the `devices` since `sensorTypes.json` is a level up (i.e., `node_modules/@zwave-js/config/config/**`) but that didn't change anything.

I also tried running the build using `yarn pkg` instead of `./package.sh`, but again no difference in behavior.

**Reply by AlCalzone** [reply]
Does `**` include files at the current level? In some environments it only includes folders and their contents.

**Reply by justindthomas** [reply]
I attempted to widen it using:

~~~~~
"node_modules/@zwave-js/config/config/**/*",
"node_modules/@zwave-js/config/config/*"
~~~~~

...in the `assets` of `pkg` in `package.json` and then doing a full rebuild and re-package, but it doesn't seem to have any effect on that error message. It's certainly possible I'm missing a step - I'm not very familiar with `yarn`.

~~~~~
    "assets": [
      "views/**/*",
      "static/**/*",
      "dist/**/*",
      "node_modules/@serialport/bindings/**",
      "node_modules/leveldown/**",
      "node_modules/@zwave-js/config/config/**/*",
      "node_modules/@zwave-js/config/config/*"
    ],
~~~~~

~~~~~
./package.sh
~~~~~

Rebuild: yes
Platform: arm64

~~~~~
cp pkg/zwavejs2mqtt /srv/zwave/
cd /srv/zwave
./zwavejs2mqtt
~~~~~

**Reply by AlCalzone** [reply]
I'm not familiar with the packaging stuff, so I can only raise questions here with my limited knowledge:
> node_modules/@zwave-js/config/config/*

Does this include files with extensions? Or would the path have to be `node_modules/@zwave-js/config/config/*.*`?

**Reply by justindthomas** [reply]
I'm pretty sure that `*` includes all extensions based on what I read in https://github.com/isaacs/node-glob

Just now I tried including that file it references directly(`node_modules/@zwave-js/config/config/sensorTypes.json`) in that asset list in `package.json`. That didn't alter the behavior at all - it seems like that list just isn't doing anything.

**Reply by AlCalzone** [reply]
@robertsLando any idea?
@jmgiaever or you based on your snap experience?

**Reply by jmgiaever** [reply]
It works fine in snap. We use the external directory.

**Reply by robertsLando** [reply]
@justindthomas Try to replace 

```
"node_modules/@zwave-js/config/config/**/*",
      "node_modules/@zwave-js/config/config/*"
```

With 

```
"node_modules/@zwave-js/config/**"
```

**Reply by justindthomas** [reply]
@robertsLando I can try that. I did just run the `pkg` command separately using the parameters supplied in the `package.sh` script but added the debug `-d` flag and captured the output. I do see files in that directory being added.

~~~~~
> [debug] Content of /ramdisk/code/zwavejs2mqtt/node_modules/@zwave-js/config/config/sensorTypes.json is added to queue. It was required from /ramdisk/code/zwavejs2mqtt/package.json
> [debug] Stat info of /ramdisk/code/zwavejs2mqtt/node_modules/@zwave-js/config/config/sensorTypes.json is added to queue.
> [debug] Content of /ramdisk/code/zwavejs2mqtt/node_modules/@zwave-js/config/config/sensorTypes.json is added to queue. It was required from /ramdisk/code/zwavejs2mqtt/node_modules/@zwave-js/config/package.json
> [debug] Content of /ramdisk/code/zwavejs2mqtt/node_modules/@zwave-js/config/config/sensorTypes.json is added to queue. It was required from /ramdisk/code/zwavejs2mqtt/node_modules/@zwave-js/config/package.json
  /ramdisk/code/zwavejs2mqtt/node_modules/@zwave-js/config/config/sensorTypes.json
  /ramdisk/code/zwavejs2mqtt/node_modules/@zwave-js/config/config/sensorTypes.json
  sensorTypes.json
~~~~~

Could the way it's being referenced as `/snapshot/` be the problem? That seems unusual and I'm not sure how that's supposed to work - I assume that's part of the `pkg` functionality?

~~~~~
2022-01-14T16:20:07.107Z CONFIG   Synchronizing external config dir failed - using embedded config. Error: Error
                                  : ENOENT: no such file or directory, copyfile '/snapshot/zwavejs2mqtt/node_mod
                                  ules/@zwave-js/config/config/sensorTypes.json' -> '/srv/zwave/.config-db/senso
                                  rTypes.json'
~~~~~

**Reply by robertsLando** [reply]
> Could the way it's being referenced as /snapshot/ be the problem

That's correct, when you create the package it uses a snapshot file system to serve the files inside the package executable. Try to add that change in package.json and let me know, I will add that too if it works for you

**Reply by justindthomas** [reply]
I had the idea to create that structure (`/snapshot/zwavejs2mqtt/node_modules/@zwave-js/config/config`) and populate that from the directory in my code folder. That worked.

~~~~~
  481  sudo mkdir -p /snapshot/zwavejs2mqtt/node_modules/@zwave-js
  485  sudo chown -R ubuntu:ubuntu /snapshot
  486  cp -ar node_modules/@zwave-js/config /snapshot/zwavejs2mqtt/node_modules/@zwave-js/
~~~~~

~~~~~
2022-01-14T16:31:39.904Z CONFIG   Synchronizing external config dir /srv/zwave/.config-db...
2022-01-14T16:31:44.538Z CONFIG   Using external configuration dir /srv/zwave/.config-db
2022-01-14T16:31:44.539Z CONFIG   version 8.10.2
2022-01-14T16:31:52.608Z CONFIG   Device index regenerated
~~~~~

I had been assuming that `snapshot` was a special reference to pull from within the package file, but I guess not? Maybe that's something that needs to be coded in to the `package.sh` script to move those files to a predictable location. But that would break portability (i.e., the executable would only be suitable to run on the system where it was built). Hmm.

**Reply by justindthomas** [reply]
Okay, I'm rebuilding now with that change to `package.json`. Will let you know shortly.

I also used `DEBUG_PKG=1` to list the contents of the executable. It does have a  `@zwave-js` folder, but does not include the `config` directory. Hopefully this change will fix it.

**Reply by justindthomas** [reply]
That change worked to get the files in to the `snapshot` folder, but the error persists.

I went searching to see if maybe there's a problem with `fs.copy` and snapshot files and found this (which it seems you are also aware of :) ): https://github.com/vercel/pkg/issues/420

Could that be contributing?

**Reply by robertsLando** [reply]
> I went searching to see if maybe there's a problem with fs.copy and snapshot files and found this (which it seems you are also aware of :) ): vercel/pkg#420

It could be related to that if @AlCalzone is using `fs.copy` to copy files from the config folder to the external db, didn't think about that :( 

> It does have a @zwave-js folder, but does not include the config directory. Hopefully this change will fix it.

Does even adding `"node_modules/@zwave-js/config/config/**"` to package.json assets not working?

**Reply by AlCalzone** [reply]
Yeah, we're copying files. Pretty sure `fs-extra` uses `fs.copy` internally.

**Reply by robertsLando** [reply]
Humm so I think the only solution would be to apply a patch or fix the bug my own on the repo 😆

**Reply by robertsLando** [reply]
https://github.com/vercel/pkg/pull/1484/files

**Reply by justindthomas** [reply]
@robertsLando Awesome - I suspect that will help.

With regard to the `assets` inclusion, I think it may be fine the way it is. Now that I know where to look, I can't replicate what I thought I was seeing earlier with the `config` directory not being included - even when I reset to the `master` version of `package.json` and rebuild with `DEBUG` turned on, that directory seems to be included. I think it may just be that `pkg` issue that's holding things up.

**Reply by robertsLando** [reply]
I didn't do it before just because I would like to have more people contributing to that project but most times seems creating patches is easier then creating a PR 🤷🏼‍♂️ anyway let's see with next update how it goes
